// @cometjs/core/lib/option.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.isSome = isSome;
exports.isNone = isNone;
exports.optionToString = optionToString;
exports.matchOption = matchOption;
exports.mapOption = mapOption;

var _map = require("./map");

function isSome(option) {
  return option != null;
}

function isNone(option) {
  return option == null;
}

function optionToString(option) {
  return isSome(option) ? `Some(${typeof option})` : 'None';
}

function matchOption(option) {
  return isSome(option) ? 'some' : 'none';
}

function mapOption(option, map) {
  if (typeof map === 'function') {
    return isSome(option) ? map(option) : null;
  }

  if (typeof map !== 'object') {
    throw new Error(`The second argument only allows function or object but got: ${typeof map}`);
  }

  const matchedMap = map[matchOption(option)];

  if (!matchedMap) {
    throw new Error(`The object doesn't have map to ${optionToString(option)} type`);
  }

  return isSome(option) ? (0, _map.mapToValue)(matchedMap, option) : (0, _map.mapToValue)(matchedMap);
}
